libs/ml/include/ml/core/graph.hpp

```diff
@@ -401,6 +420,18 @@ TensorType Graph<TensorType>::ForwardPropagate(std::string const &node_name, boo
   return ForwardImplementation(node_name, is_training, false);
 }
 
+template <typename TensorType>
```

Comment:
this stuff is really not simple at all and **wont** be obvious when we look back at it in the future so please add a doxygen-style function comment explaining **which** node node_name is pointing to when you invoke this function and a description of what the function is for more generally

Reply:
added a comment